### ISSUE TITLE
[ENH] Keeping a history of every operation applied to a DataFrame

### ISSUE THREAD

**Issue by zbarry** [ericmjl/pyjanitor#475]
A whacky idea dreamed up during the sprint.

This would essentially record for you your data processing pipeline (on the DataFrame bit at least). Metadata could go in a `.pj` accessor in the dataframe / series.

Fun challenges include:

* Detecting use of all Pandas functions, not just PyJanitor ones
* Minimizing modification of Pandas objects as much as possible
* Avoiding brittleness to pandas code updates
* Handling multi-dataframe operations without losing computation metadata

Thoughts on how this could be possible would be nice. Dream big.

@szuckerman @ericmjl @HectorM14

**Comment by jcmkk3** [comment]
This is not exactly the same thing, but since you’re already thinking about a scope outside of just pyjanitor added methods, it would be really cool to see something like [tidylog](https://github.com/elbersb/tidylog/blob/master/README.md) for Pandas.

**Comment by ericmjl** [comment]
@jcmkk3 tidylog is a very good idea! A hacky first-pass would be have every `pyjanitor` function print out what exactly it’s doing. This, of course, ignores all `pandas` functions. The more sophisticated version would use the logging facilities in Python. `pyjanitor` functions would need to be changed to explicitly report what exactly it’s doing in plain English; with an additional standard kwarg, we could allow a user to provide a reason for doing something, passed in as a string to the function call. (It may look clunky, though.) 

Anyways, just some ideas to riff off @jcmkk3’s idea.

**Comment by szuckerman** [comment]
Apparently you can put a trace function in python that will log every function that's called. I saw [this answer](https://stackoverflow.com/a/8315566) on stackoverflow which gave some decent code to show which functions are being called.

However, one of the main problems is how many intermediate functions are getting called. I modified the code from the stackoverflow answer to write all the results to a list.

```python
import sys
import os

log = []

def tracefunc(frame, event, arg, indent=[0]):
      if event == "call":
          log.append(frame.f_code.co_name)
      return tracefunc

# sys.settrace(tracefunc)
sys.setprofile(tracefunc)

def main():
	print(os.getcwd())

main() 
log
```

If I run the above, the output is:

```python
['<module>', '<module>', '<module>', 'main', '<module>']
```

Now if I just merely import the pandas module:

```python
import pandas as pd
len(log)
>>> 141765
```

And if I make a DataFrame:

```python
dat = pd.DataFrame({'name': ['first1', 'first2'], 'last': ['last1', 'last2']})
len(log)
>>> 142220
```

Just creating a DataFrame was 455 function calls.

And just calling a DataFrame adds nearly 10k function calls.

```python
dat
len(log)
>>> 153782
```

However, if you log[-10:], the output is:

```python
['justify', 'justify', '<listcomp>', '_join_unicode', '_join_unicode', '_join_unicode', '_join_unicode', 'should_show_dimensions', '<module>', '<module>']
```

I'm thinking that if in the `tracefunc` function there was a condition to check for the set of all possible DataFrame and PyJanitor methods it would cut the list down to only what we want. Also, one thing to think about is how the DataFrame methods within PyJanitor functions will be logged, since they're getting called as well.

**Comment by hectormz** [comment]
Interesting idea!

 Would the goal be to first log operations for a dataframe, and be able reload those operations, or generate a `.py` with the same workflow? Or as @ericmjl says, a report in plain English. And should we be able to account for the possibility of multiple dataframes being processed, instead of recording all `pyjanitor` and `pandas` functions?

An alternative to some type of logging could be a decorator that gets applied to all `pandas`/`pyjanitor` methods. I haven't tried it out, but this [Stack Overflow post (Defining Python decorators for a complete module)](https://stackoverflow.com/questions/8951787/defining-python-decorators-for-a-complete-module) has some example code:

```python
import othermodule
import types
for k,v in vars(othermodule).items():
    if isinstance(v, types.FunctionType):
        vars(othermodule)[k] = my_decorator(v)
```

That code (and more) could be put into a method like `janitor.record()`, invoked at the top of a script or notebook.

It may be helpful to see some examples of `pandas`/`pyjanitor` operations, and the resulting records, regardless of how we get there.

**Comment by szuckerman** [comment]
@hectormz I would vote for @ericmjl's suggestion about having the report in English. Reloading the operations would be nice, but users already have the code in a notebook or .py file; I'm not sure as to how helpful a log in that case would be.

I would vote for the following: Have a log like a web-style log. It's basically a DataFrame with columns like:

`datetime`, `function_call`, `meaning`, `current_shape_of_df`, `time_to_execute`

and maybe also columns for system settings (operating system, python version, etc).

What makes this difficult, though, is that to make it useful a config file would probably be necessary.

Some initial questions that come to mind that could be solved by a config would be stuff like:
1. Where is this being output? a `DataFrame`? a `file'? a `sqlite` db? 
1. How many records to store before removing values? Maybe removing after a certain amount of time? A new python session?
1. How verbose should the logging be? Only PyJanitor? Also Pandas? Also base Python functions?

This all sounds like interesting stuff that users would appreciate, but I'm wondering if it's too big a task for this package. Maybe make this its own package that works with PyJanitor?

**Comment by hectormz** [comment]
Now I understand, got it. Beyond the implementation of the history logging, one of the big questions to me will be where to store the history (and for how long), if a config file is not set.

In my view, the big picture items are:

1. Actual implementation of `pandas`/`pyjanitor` logging
2. Creating dictionary of methods and plain English equivalents (manually or automatically)
    - Maybe creating a small tool to check codebases for methods without translations as the `pyjanitor`/`pandas` methods change
3. Formatting and saving output/managing file size

It seems like a majority of effort is packed into 1., but does that cover the basic considerations?

`tidylog` looks very nice @jcmkk3 !

**Comment by ericmjl** [comment]
Friends, I think @eyaltrabelsi, a contributor to pyjanitor, has done some wonderful work with [pandas-log](https://github.com/eyaltrabelsi/pandas-log) related to this issue! The programming  model is very good: context manager so that we can selectively do logging at the appropriate places.

**Comment by ericmjl** [comment]
I am going to close this issue for now, but please feel free to re-open later!
